presto-hive/src/main/java/com/facebook/presto/hive/CopyOnWriteConfiguration.java

```diff
@@ -0,0 +1,465 @@
+    }
+
+    @Override
+    public Class<?>[] getClasses(String name, Class<?>... defaultValue)
+    {
+        return config.getClasses(name, defaultValue);
+    }
+
+    @Override
+    public Class<?> getClass(String name, Class<?> defaultValue)
+    {
+        return config.getClass(name, defaultValue);
+    }
+
+    @Override
+    public <U> Class<? extends U> getClass(String name, Class<? extends U> defaultValue, Class<U> xface)
+    {
+        return config.getClass(name, defaultValue, xface);
+    }
+
+    @Override
+    public <U> List<U> getInstances(String name, Class<U> xface)
+    {
+        return config.getInstances(name, xface);
+    }
+
+    @Override
+    public Path getLocalPath(String dirsProp, String path)
+            throws IOException
+    {
+        return config.getLocalPath(dirsProp, path);
+    }
+
+    @Override
+    public File getFile(String dirsProp, String path)
+            throws IOException
+    {
+        return config.getFile(dirsProp, path);
+    }
+
+    @Override
+    public URL getResource(String name)
+    {
+        return config.getResource(name);
+    }
+
+    @Override
+    public InputStream getConfResourceAsInputStream(String name)
+    {
+        return config.getConfResourceAsInputStream(name);
+    }
+
+    @Override
+    public Reader getConfResourceAsReader(String name)
+    {
+        return config.getConfResourceAsReader(name);
+    }
+
+    @Override
+    public Set<String> getFinalParameters()
+    {
+        return config.getFinalParameters();
+    }
+
+    @Override
+    public int size()
+    {
+        return config.size();
+    }
+
+    @Override
+    public void clear()
+    {
+        config.clear();
+    }
+
+    @Override
+    public Iterator<Map.Entry<String, String>> iterator()
+    {
+        return config.iterator();
+    }
+
+    @Override
+    public ClassLoader getClassLoader()
+    {
+        return config.getClassLoader();
+    }
+
+    @Override
+    public void readFields(DataInput in)
+            throws IOException
+    {
+        config.readFields(in);
+    }
+
+    @Override
+    public void write(DataOutput out)
+            throws IOException
+    {
+        config.write(out);
+    }
+
+    @Override
+    public void writeXml(OutputStream out)
+            throws IOException
+    {
+        config.writeXml(out);
+    }
+
+    @Override
+    public void writeXml(Writer out)
+            throws IOException
+    {
+        config.writeXml(out);
+    }
+
+    @Override
+    public Map<String, String> getValByRegex(String regex)
+    {
+        return config.getValByRegex(regex);
+    }
+
+    @Override
+    public String toString()
+    {
+        return config.toString();
+    }
+
+    private void checkAndSet(Runnable action)
+    {
+        if (!isMutable) {
+            Configuration originalConfig = config;
+            config = new Configuration(originalConfig);
+            isMutable = true;
+        }
+        action.run();
```

Comment:
Will this set the old config or the new config? Maybe the old config is captured as part of Runnable? Please add a unit test to make sure the old config is not modified.